tests/storage/zfsvol/test_zfsvol_sr.py

```diff
@@ -0,0 +1,72 @@
+import logging
+import time
+import pytest
+
+from lib.commands import SSHCommandFailed
+from lib.common import wait_for, vm_image
+from tests.storage import vdi_is_open
+
+# Requirements:
+# - one XCP-ng host >= 8.2 with an additional unused disk for the SR
+# - access to XCP-ng RPM repository from the host
+
+@pytest.mark.usefixtures("sr_disk_wiped")
+class TestZfsvolSRCreateDestroy:
+    """
+    Tests that do not use fixtures that setup the SR or import VMs,
+    because they precisely need to test SR creation and destruction,
+    and VM import.
+    """
+
+    def test_create_and_destroy_sr(self, host, sr_disk_wiped):
+        # Create and destroy tested in the same test to leave the host as unchanged as possible
+        sr = host.sr_create('zfs-vol', "ZFS-local-SR-test", {'device': '/dev/' + sr_disk_wiped}, verify=True)
+        # import a VM in order to detect vm import issues here rather than in the vm_on_xfs_fixture used in
+        # the next tests, because errors in fixtures break teardown
+        vm = host.import_vm(vm_image('mini-linux-x86_64-bios'), sr_uuid=sr.uuid)
+        vm.destroy(verify=True)
+        sr.destroy(verify=True)
+
+@pytest.mark.usefixtures("zfsvol_sr")
+class TestZfsvolSR:
+
+    @pytest.mark.xfail
+    @pytest.mark.quicktest
+    def test_quicktest(self, zfsvol_sr):
+        zfsvol_sr.run_quicktest()
+
+    @pytest.mark.small_vm # run with a small VM to test the features
+    @pytest.mark.big_vm # and ideally with a big VM to test it scales
+    def test_start_and_shutdown_VM(self, vm_on_zfsvol_sr):
+        vm = vm_on_zfsvol_sr
+        vm.start()
+        vm.wait_for_os_booted()
+        vm.shutdown(verify=True)
+
+    @pytest.mark.xfail # needs support for destroying snapshots
+    @pytest.mark.small_vm
+    @pytest.mark.big_vm
+    def test_snapshot(self, vm_on_zfsvol_sr):
+        vm = vm_on_zfsvol_sr
+        vm.start()
+        try:
+            vm.wait_for_os_booted()
+            vm.test_snapshot_on_running_vm()
+        finally:
+            vm.shutdown(verify=True)
+
+    # *** tests with reboots (longer tests).
+
+    @pytest.mark.reboot
+    @pytest.mark.small_vm
+    def test_reboot(self, vm_on_zfsvol_sr, host, zfsvol_sr):
+        sr = zfsvol_sr
+        vm = vm_on_zfsvol_sr
+        host.reboot(verify=True)
+        wait_for(sr.all_pbds_attached, "Wait for PBD attached")
```

Comment:
when testing that zpool is mounted after reboot I observed that it needs several minutes to be mounted. I mean once rebooted I logged into the host and running `zpool list` show me nothing and xapi was already started. So it seems that there is a few delay to be mounted. But wait_for is waiting 2 minutes so it should be ok but if this test failed we need to check if it is not an issue with timeout.

Reply:
> It's best to investigate why such a delay can happen, but if really needed `wait_for` can accept a different timeout value.

It is because after a reboot the zpool will be mounted when SR.attach is called the first time so it is not it is not necessarily at the very beginning of starting the xapi.

Reply:
And what calls SR.attach? For all existing smapiv1 drivers, PBDs are plugged directly after boot so that the SR is immediately available to VMs.

Reply:
It can be called at different places. If the PBD is plugged then SR.attach() has been called. PBDs are not plugged "directly" after the boot. There is the initialization of the XAPI itself that happens after systemd has started some services required by the XAPI. So after booting you have some delay that can be one or 2 minutes. For example I can ssh to the host while XAPI has not yet started... So ssh is started before the XAPI.

Reply:
Yes, `SR.attach` is the implementation of `xe pbd-plug`, the kind of stuff that XAPI will run when it starts up to bring the system up in sync with its db.

Reply:
In the tests, we wait for XAPI to be available, and PBDs are then plugged automatically, so I don't understand why we reach a timeout specifically for this SR driver and not others.

Reply:
What I see is a rather unreasonable waiting time to notice the host is back up after reboot:
```
Apr 02 16:24:48 INFO Wait for host up
Apr 02 16:29:49 INFO Wait for PBD attached
Apr 02 16:29:50 INFO Start VM
```

It seems like `ssh` stays stuck quite some time when launched while the IP is not reachable yet.  If we check the IP with `ping` first, the waiting time is much reduced:

```
Apr 02 16:47:34 INFO Wait for host down
Apr 02 16:47:38 INFO Wait for host up
Apr 02 16:47:38 INFO Wait for XAPI to be ready
Apr 02 16:49:10 INFO Wait for PBD attached
```

Will submit a separate PR for this.

Reply:
In fact there are cases where the nested host under test takes much more time to boot (can be see in the console in XO):
```
Apr 02 17:24:48 INFO Wait for host down
Apr 02 17:24:49 INFO Wait for ssh up on host
Apr 02 17:25:45 INFO Wait for XAPI to be ready
Apr 02 17:26:22 INFO Wait for PBD attached
```

So likely "not the fault of anything SMAPIv3"

Reply:
And maybe the root cause of the delay is not so obvious, or not single:
```
Apr 02 17:32:18 INFO Wait for host down
Apr 02 17:32:22 INFO Wait for host up
Apr 02 17:32:22 INFO Wait for ssh up on host
Apr 02 17:33:10 INFO Wait for XAPI to be ready
Apr 02 17:33:50 INFO Wait for PBD attached
```

Reply:
submitted #219 for that specific issue